Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp

```diff
@@ -243,13 +372,71 @@ void KafkaEventStreamDecoder::eventDataFromMessage(const std::string &buffer) {
   } else {
     periodBuffer = m_localEvents[0];
   }
-  for (decltype(nEvents) i = 0; i < nEvents; ++i) {
-    auto &spectrum = periodBuffer->getSpectrum(
-        m_specToIdx[static_cast<int32_t>(detData[i])]);
-    spectrum.addEventQuickly(TofEvent(static_cast<double>(tofData[i]) *
-                                          1e-3, // nanoseconds to microseconds
-                                      pulseTime));
+
+  auto numGroups = tofData.size() / MAX_BUFFERGROUP_EVENTS;
+  std::vector<size_t> sizes(numGroups + 1, MAX_BUFFERGROUP_EVENTS);
+  sizes[numGroups] = tofData.size() % MAX_BUFFERGROUP_EVENTS;
+  size_t start = 0;
+
+  auto starttime = std::chrono::system_clock::now();
+  for (auto &grpSize : sizes) {
+    if (grpSize == 0)
+      continue;
+
+    BufferGroup buffGrp;
+    buffGrp.tof.resize(grpSize);
+    buffGrp.detID.resize(grpSize);
+
+    auto end = start + grpSize;
+    std::copy(tofData.data() + start, tofData.data() + end,
+              buffGrp.tof.begin());
+    std::copy(detData.data() + start, detData.data() + end,
+              buffGrp.detID.begin());
+
+    buffGrp.pulseTime = pulseTime;
+
+    buffGrp.periodBuffer = periodBuffer;
+    m_buffer.emplace_back(buffGrp);
+    start += grpSize;
+  }
+
+  auto endTime = std::chrono::system_clock::now();
+  std::chrono::duration<double> dur = endTime - starttime;
+  totalEventFromMessageDuration += dur.count();
+  numEventFromMessageCalls += 1;
+}
+
+void KafkaEventStreamDecoder::populateEventDataWorkspace() {
+  auto startTime = std::chrono::system_clock::now();
+  for (auto &grp : m_buffer) {
+    auto nEvents = grp.tof.size();
+
+    if (nEvents == 0)
+      continue;
+
+    auto specindices = getSpectrumIndices(grp.detID.data(), nEvents);
+    auto perm = getSortedSpectraPermutation(specindices);
+    auto ranges = getRanges(perm, specindices);
+    PARALLEL_SET_NUM_THREADS(16)
+    PARALLEL_FOR_NO_WSP_CHECK()
+    for (int r = 0; r < ranges.size(); r++) {
+      auto &range = ranges[r];
+      for (int j = range.first; j < range.second; j++) {
+        auto i = perm[j];
+        auto &spectrum = grp.periodBuffer->getSpectrum(specindices[i]);
+        spectrum.addEventQuickly(
+            TofEvent(static_cast<double>(grp.tof[static_cast<uint32_t>(i)]) *
+                         1e-3, // nanoseconds to microseconds
+                     grp.pulseTime));
+      }
+    }
   }
+  m_buffer.clear();
```

Comment:
In addition to the buffer for event data which should be reused (see previous comments), `m_buffers` also contains a `shared_ptr` to `EventWorkspace`. Constantly creating and destroying those is also relatively costly and should be avoided. Can they be reused? Otherwise, can we use a raw pointer (as far as I can tell `m_buffer` does not outlive the other owners since it is just used internally)?